<launch>

  <!-- simulation args -->
  <arg name="gazebo_gui"  default="false" />
  <arg name="rviz"        default="true" />
  <arg name="network_viz" default="true"/>

  <!-- patrol demo args -->
  <arg name="task_agent_velocity"    default="1.0"/>  <!-- m/s -->
  <arg name="network_agent_velocity" default="2.0"/>  <!-- m/s -->
  <arg name="sample_count"           default="5000"/>
  <arg name="sample_variance"        default="2.0"/>
  <arg name="margin"                 default="0.15"/> <!-- [0 1] -->
  <arg name="confidence"             default="0.70"/> <!-- (0.5 1) -->
  <arg name="collision_distance"     default="4.0"/>  <!-- m -->
  <arg name="heading_tol"            default="3.5"/>  <!-- > pi : agents don't turn before moving -->
  <arg name="minimum_update_rate"    default="10"/>   <!-- Hz -->
  <arg name="patrol_radius"          default="20.0"/> <!-- radius of patrol circle -->

  <!-- patrol demo params -->
  <rosparam param="/task_agent_ids">[1, 2, 3]</rosparam> <!-- needed for task agent count -->
  <rosparam param="/comm_agent_ids">[4, 5, 6]</rosparam> <!-- needed for network agent count -->
  <param name="/patrol_radius"       value="$(arg patrol_radius)"/>
  <param name="/desired_altitude"    value="2.0"/>

  <!-- channel rate model params -->
  <param name="/N0" value="-70.0"/> <!-- -85.0  receiver noise (dBm) -->
  <param name="/n"  value="2.52"/>  <!--   2.52 decay exponent -->
  <param name="/L0" value="-48.0"/> <!-- -53.0  transmit power (dBm) -->
  <param name="/a"  value="0.2"/>   <!--   0.2  variance sigmoid param 1 -->
  <param name="/b"  value="0.6"/>   <!--   0.6  variance sigmoid param 2 -->

  <!-- gazebo and rviz -->
  <include file="$(find intel_aero_gazebo)/launch/simulation.launch">
    <arg name="gazebo_gui"       value="$(arg gazebo_gui)"/>
    <arg name="world"            value="worlds/empty.world"/>
    <arg name="rviz"             value="$(arg rviz)"/>
    <arg name="rviz_config_file" value="$(find intel_aero_demos)/rviz/patrol.rviz"/>
  </include>

  <!-- load saved map -->
  <node pkg="map_server"
        type="map_server"
        name="static_map_node"
        args="$(find intel_aero_gazebo)/maps/empty.yaml">
    <remap from="map"      to="/projected_map"/>
    <param name="frame_id" value="world"/>
  </node>

  <!-- network topology visualization -->
  <node pkg="network_status"
        type="routing_visualization_node"
        name="routing_visualization_node"
        output="screen">
    <param name="world_frame" value="world"/>
    <param name="pose_topic"  value="pose"/>
  </node>

  <!-- network planner nodes -->

  <node pkg="network_planner"
        type="network_planner_patrol_node"
        name="network_planner_patrol_node"
        output="screen">
    <param name="sample_count"          value="$(arg sample_count)"/>
    <param name="sample_variance"       value="$(arg sample_variance)"/>
    <param name="max_velocity"          value="$(arg task_agent_velocity)"/>
    <param name="margin"                value="$(arg margin)"/>
    <param name="confidence"            value="$(arg confidence)"/>
    <param name="collision_distance"    value="$(arg collision_distance)"/>
    <param name="minimum_update_rate"   value="$(arg minimum_update_rate)"/>
    <param name="pose_topic"            value="pose"/>
    <param name="nav_nodelet"           value="gazebo_vel_nav_nodelet"/>
    <param name="world_frame"           value="world"/>
  </node>

  <node pkg="socp"
        type="rr_socp_server.py"
        name="robust_routing_socp_server"
        output="screen"/>

  <!-- task (patrol) agents -->

  <group>
    <arg name="id" value="1"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="$(eval arg('patrol_radius')*cos(2.0*3.141592/3.0*(arg('id')-1)))"/>
      <arg name="y"              value="$(eval arg('patrol_radius')*sin(2.0*3.141592/3.0*(arg('id')-1)))"/>
      <arg name="linear_vel_des" value="$(arg task_agent_velocity)"/>
      <arg name="heading_tol"    value="$(arg heading_tol)"/>
    </include>
  </group>

  <group>
    <arg name="id" value="2"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="$(eval arg('patrol_radius')*cos(2.0*3.141592/3.0*(arg('id')-1)))"/>
      <arg name="y"              value="$(eval arg('patrol_radius')*sin(2.0*3.141592/3.0*(arg('id')-1)))"/>
      <arg name="linear_vel_des" value="$(arg task_agent_velocity)"/>
      <arg name="heading_tol"    value="$(arg heading_tol)"/>
    </include>
  </group>

  <group>
    <arg name="id" value="3"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="$(eval arg('patrol_radius')*cos(2.0*3.141592/3.0*(arg('id')-1)))"/>
      <arg name="y"              value="$(eval arg('patrol_radius')*sin(2.0*3.141592/3.0*(arg('id')-1)))"/>
      <arg name="linear_vel_des" value="$(arg task_agent_velocity)"/>
      <arg name="heading_tol"    value="$(arg heading_tol)"/>
    </include>
  </group>

  <!-- network agents -->

  <arg name="net_rad" default="8.0"/>

  <group>
    <arg name="id" value="4"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="$(eval arg('net_rad')*cos(2.0*3.141592/3.0*(arg('id')-4)))"/>
      <arg name="y"              value="$(eval arg('net_rad')*sin(2.0*3.141592/3.0*(arg('id')-4)))"/>
      <arg name="linear_vel_des" value="$(arg network_agent_velocity)"/>
      <arg name="heading_tol"    value="$(arg heading_tol)"/>
    </include>
  </group>

  <group>
    <arg name="id" value="5"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="$(eval arg('net_rad')*cos(2.0*3.141592/3.0*(arg('id')-4)))"/>
      <arg name="y"              value="$(eval arg('net_rad')*sin(2.0*3.141592/3.0*(arg('id')-4)))"/>
      <arg name="linear_vel_des" value="$(arg network_agent_velocity)"/>
      <arg name="heading_tol"    value="$(arg heading_tol)"/>
    </include>
  </group>

  <group>
    <arg name="id" value="6"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="$(eval arg('net_rad')*cos(2.0*3.141592/3.0*(arg('id')-4)))"/>
      <arg name="y"              value="$(eval arg('net_rad')*sin(2.0*3.141592/3.0*(arg('id')-4)))"/>
      <arg name="linear_vel_des" value="$(arg network_agent_velocity)"/>
      <arg name="heading_tol"    value="$(arg heading_tol)"/>
    </include>
  </group>

</launch>
